Update PropertyValueConverter examples to Umbraco 15+

Luuk1983 opened the pull request:
## 📋 Description
This PR updates the examples of PropertyValueConverters to Umbraco 15+. I updated both the Umbraco 15 and 16 documentation. And while I was at it, I also revised the pages thoroughly.

## 📎 Related Issues (if applicable)
Fixes [#7336](https://github.com/umbraco/UmbracoDocs/issues/7336)

## ✅ Contributor Checklist

I've followed the [Umbraco Documentation Style Guide](https://docs.umbraco.com/contributing/documentation/style-guide) and can confirm that:

* [x] Code blocks are correctly formatted.
* [x] Sentences are short and clear (preferably under 25 words).
* [x] Passive voice and first-person language (“we”, “I”) are avoided.
* [x] Relevant pages are linked.
* [x] All links work and point to the correct resources.
* [x] Screenshots or diagrams are included if useful.
* [x] Any code examples or instructions have been tested.
* [x] Typos, broken links, and broken images are fixed.

## Product & Version (if relevant)
v15 and v16 (versions with the new cache)

## Deadline (if relevant)
-

## 📚 Helpful Resources

* 🧾 [Umbraco Contribution Guidelines](https://docs.umbraco.com/contributing)
* ✍️ [Umbraco Documentation Style Guide](https://docs.umbraco.com/contributing/documentation/style-guide)

mikecp reviewed (commented):
Great rewrite @Luuk1983, all very clear! I added a few comments but nothing fundamental, mainly small stuff for keeping consistency throughout the whole text and such.

I hope I did not go too far into details  😅

Cheers!

Luuk1983 commented on 15/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -4,61 +4,37 @@ description: A guide to creating a custom property value converter in Umbraco
+A Property Value Converter converts a property editor's database-stored value into another type that is stored in the Umbraco cache. This way, the database stores only the most essential data, while Razor views, the Published Content API, and the Content Delivery API can work with strongly typed and cleaner models.
 
-Published property values have four "Values":
+For example, a Document picker typically only stores the Key of the picked node in the database, but when working with the published data, an IPublishedContent object is returned instead of just the key. This conversion is done by a Property Value Converter.
```
I made a pass to make it consistent

Luuk1983 commented on 15/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -4,61 +4,37 @@ description: A guide to creating a custom property value converter in Umbraco
-* **Source** - The raw data stored in the database, this is generally a `String`
-* **Intermediate** - An object of a type that is appropriate to the property, for example a nodeId should be an `Int` or a collection of nodeIds would be an integer array, `Int[]`
-* **Object** - The object to be used when accessing the property using a Published Content API, for example UmbracoHelper's `GetPropertyValue<T>` method
+A PropertyValueConverter has three conversion levels:
```
I agree, I made a pass to make it consistent

Luuk1983 commented on 15/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -4,61 +4,37 @@ description: A guide to creating a custom property value converter in Umbraco
-If you are implementing a PropertyValueConverter for a PropertyEditor that doesn't already have one, creating the PropertyValueConverter will automatically enable it. No further actions are needed.
-
-If you aim to override an existing PropertyValueConverter, possibly from Umbraco or a package, additional steps are necessary. Deregister the existing one to prevent conflicts in this scenario.
+## Create a PropertyValueConverter
```
I made a pass to make it consistent

Luuk1983 commented on 15/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -4,61 +4,37 @@ description: A guide to creating a custom property value converter in Umbraco
-* **Object** - The object to be used when accessing the property using a Published Content API, for example UmbracoHelper's `GetPropertyValue<T>` method
+A PropertyValueConverter has three conversion levels:
+* **Source** - The raw data stored in the database; this is generally a `string`.
+* **Intermediate** - An object of a type that is appropriate to the property. For example, a node key should be a `Guid`, or a collection of node keys would be a `Guid[]` array.
```
Fixed!

Luuk1983 commented on 16/umbraco-cms/.gitbook.yaml:
```diff
@@ -113,6 +113,7 @@ redirects:
     reference/configuration/runtimeminificationsettings: reference/configuration/README.md
     reference/routing/umbraco-api-controllers/authorization: reference/routing/umbraco-api-controllers/README.md
     reference/routing/umbraco-api-controllers/routing: reference/routing/umbraco-api-controllers/README.md
+    reference/notifications/sendingallowedchildrennotifications: reference/content-type-filters.md
```
This is a leftover of a previous PR, my bad. It's gone now!

Luuk1983 commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
They are EXACTLY the same, so yeah, they apply to both V15 and V16 ;)

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -1,62 +1,37 @@
+## Implement information methods
+Implement the following methods, which provide Umbraco with context about the Property Value Converter.
 
 ### IsConverter(IPublishedPropertyType propertyType)
```
```suggestion
### `IsConverter(IPublishedPropertyType propertyType)`
```

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -1,62 +1,37 @@
-This method is called for each PublishedPropertyType (Document Type Property) at application startup. By returning `True` your value converter will be registered for that property type and your conversion methods will be executed whenever that value is requested.
+This method is called for each PublishedPropertyType (Document Type Property) at Umbraco startup. By returning `true`, your value converter will be registered for that property type. Umbraco then executes your conversion methods whenever that value is requested.
 
 Example: Checking if the IPublishedPropertyType EditorAlias property is equal to the alias of the core content editor.\
```
```suggestion
Example: Checking if the `IPublishedPropertyType.EditorAlias` property is equal to the alias of the core content editor.\
```

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
````diff
@@ -69,12 +44,70 @@ public bool IsConverter(IPublishedPropertyType propertyType)
 ```
 
 ### IsValue(object value, PropertyValueLevel level)
+The IsValue method determines whether a property contains a meaningful value or should be considered "empty" at different stages of the value conversion process. This method is essential for Umbraco's property.HasValue() method.
````
```suggestion
The `IsValue` method determines whether a property contains a meaningful value or should be considered "empty" at different stages of the value conversion process. This method is essential for Umbraco's `property.HasValue()` method.
```

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
````diff
@@ -69,12 +44,70 @@ public bool IsConverter(IPublishedPropertyType propertyType)
+}
+```
 
 ### GetPropertyValueType(IPublishedPropertyType propertyType)
````
```suggestion
### `GetPropertyValueType(IPublishedPropertyType propertyType)`
```

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -89,11 +122,7 @@ public Type GetPropertyValueType(IPublishedPropertyType propertyType)
 
```
GitHub won't let me add a suggestion here, but for consistency, I would suggest:

1. Format the heading as code.
2. Omit the return value, as it's not included in any other heading.

That is - I suggest it should be: 

```md
### `GetPropertyCacheLevel(IPublishedPropertyType propertyType)`
```

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -89,11 +122,7 @@ public Type GetPropertyValueType(IPublishedPropertyType propertyType)
-#### `PropertyCacheLevel.Unknown`
-
-Do not use this cache level unless you know exactly what you're doing. We recommend using the `PropertyCacheLevel.Element` level.
+A property value can be cached at the following levels. This is the most commonly used cache level and should be your default, unless you have specific reasons to do otherwise.
```
```suggestion
A property value can be cached at the following levels.
```

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -102,31 +131,23 @@ The property value will be cached until its _element_ is modified. The element i
 * For properties used at the page level, the element is the entire page.
 * For properties contained within Block List items, the element is the individual Block List item.
 
-This is the most commonly used cache level and should be your default, unless you have specific reasons to do otherwise.
```
I'm not sure why this was removed from here and inserted above. I think it does belong here (see also my other comment above).

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -135,17 +156,16 @@ public PropertyCacheLevel GetPropertyCacheLevel(IPublishedPropertyType propertyT
+Implement the methods that perform the conversion from a raw database value to an intermediate value and then to the final type. Conversions happen in two steps.
 
 ### ConvertSourceToIntermediate(IPublishedElement owner, IPublishedPropertyType propertyType, object source, bool preview)
+This method converts the raw data value into an appropriate intermediate type that is needed for the final conversion step to an object. For example, a Content Picker saves the node identifier as a `string`. To return `IPublishedContent`, the final conversion step needs a `Udi` instead. So in the intermediate step, check if the string value is a valid `Udi` and convert the string to a `Udi` as th ...
```
```suggestion
This method converts the raw data value into an appropriate intermediate type that is needed for the final conversion step to an object.

For example:
- A basic text property likely just stores its data as a `string`, so that can be converted to a `string` intermediate value.
- A Content Picker stores the node identifier (`Udi`) as a `string`. To return `IPublishedContent`, the final conversion step needs a `Udi` instead. So in the intermediate step, check if the `string` value is a valid `Udi` and convert the `string` to a `Udi` as the intermediate value.
```

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -135,17 +156,16 @@ public PropertyCacheLevel GetPropertyCacheLevel(IPublishedPropertyType propertyT
-This method should convert the raw data value into an appropriate type. For example, a node identifier stored as a `String` should be converted to an `Int` or `Udi`.
-
-Include a `using Umbraco.Extensions;` to be able to use the `TryConvertTo` extension method.
+Include `using Umbraco.Extensions` to be able to use the `TryConvertTo` extension method.
```
```suggestion
```

I would remove this. It removes focus from the essentials, and any proper IDE will automatically suggest the `using`.

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -135,17 +156,16 @@ public PropertyCacheLevel GetPropertyCacheLevel(IPublishedPropertyType propertyT
-
-Include a `using Umbraco.Extensions;` to be able to use the `TryConvertTo` extension method.
+Include `using Umbraco.Extensions` to be able to use the `TryConvertTo` extension method.
 
```
Perhaps split these examples into two code blocks for clarity?

Also I suggest using the `ConvertSourceToIntermediate` implementation from the full sample here... so something along these lines:

```csharp
// Basic text property example of intermediate conversion
public object ConvertSourceToIntermediate(IPublishedElement owner, IPublishedPropertyType propertyType, object source, bool preview)
{
    return source as string;
}
```

```csharp
// Content picker property example: Converting the source value (string) to an intermediate value (GuidUdi)
public object? ConvertSourceToIntermediate(IPublishedElement owner, IPublishedPropertyType propertyType, 
	object? source, bool preview)
{
	if (source is not string { Length: > 0 } stringValue)
		return null;
		
	return UdiParser.TryParse(stringValue, out GuidUdi? guidUdi) ? guidUdi : null;
}
```

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -160,46 +180,76 @@ public object ConvertSourceToIntermediate(IPublishedElement owner, IPublishedPro
+This method converts the intermediate value to an object of the type specified in the `GetPropertyValueType()` method of the Property Value Converter. The returned value is used by the `GetPropertyValue<T>` method of `IPublishedContent`.
 
-The below example converts the nodeId (converted to `Int` or `Udi` by _ConvertSourceToIntermediate_) into an 'IPublishedContent' object.
+The example below converts the node ID (converted to `int` or `Udi` by _ConvertSourceToIntermediate_) into an `IPublishedContent` object.
```
```suggestion
The example below converts an intermediate `Udi` value (converted from `string` to `Udi` by `ConvertSourceToIntermediate`) into an `IPublishedContent` object.
```

If we use the implementation from the full sample, this should be updated. Also see below.

kjac commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
````diff
@@ -160,46 +180,76 @@ public object ConvertSourceToIntermediate(IPublishedElement owner, IPublishedPro
+The example below converts the node ID (converted to `int` or `Udi` by _ConvertSourceToIntermediate_) into an `IPublishedContent` object.
 
 ```csharp
+// In this example _contentCache is an instance of IPublishedContentCache that is injected
````
If we use the implementation from the full sample here, replace the code below with this.

```
// Converts the intermediate value (GuidUdi) to the actual object value (IPublishedContent)
public object? ConvertIntermediateToObject(IPublishedElement owner, IPublishedPropertyType propertyType,
	PropertyCacheLevel referenceCacheLevel, object? inter, bool preview)
	=> inter is GuidUdi guidUdi
		? publishedContentCache.GetById(guidUdi.Guid)
		: null;
```

kjac reviewed (changes requested):
@Luuk1983 great work!

I have left a few comments:

Most of them are cosmetic changes (even changes to the original article in many cases).

Some have to do with the sample code. I do believe we should reuse the "full example implementation" here to the extend possible.

Note that my comments are only on the V16 version of the article. The V15 one needs updating accordingly.

Also kudos to the "full example implementation". You've actually uncovered a bug in the hybrid cache implementation 🙈 I'll make sure to report that and have it fixed; the example code is correct, the core just misbehaves a bit 🙄 

For reference, here's the issue report: https://github.com/umbraco/Umbraco-CMS/issues/20051

Luuk1983 commented:
> @Luuk1983 great work!
> 
> I have left a few comments:
> 
> Most of them are cosmetic changes (even changes to the original article in many cases).
> 
> Some have to do with the sample code. I do believe we should reuse the "full example implementation" here to the extend possible.
> 
> Note that my comments are only on the V16 version of the article. The V15 one needs updating accordingly.
> 
> Also kudos to the "full example implementation". You've actually uncovered a bug in the hybrid cache implementation 🙈 I'll make sure to report that and have it fixed; the example code is correct, the core just misbehaves a bit 🙄
> 
> For reference, here's the issue report: [umbraco/Umbraco-CMS#20051](https://github.com/umbraco/Umbraco-CMS/issues/20051)

@kjac, I updated the PR and incorporated your changes. I also updated the Umbraco 15 docs. And I see that the bug I accidentally discovered is fixed in Umbraco 16.3.0, so I guess this PR is good to go :)

eshanrnh commented on 15/umbraco-cms/customizing/property-editors/full-examples-value-converters.md:
```diff
@@ -1,6 +1,6 @@
-# Content Picker Value Converter Example
+# Property Value Converter full example
```
You’ll also need to update the Summary.md file with the new title for it to show up correctly in GitBook 😊I'll take care of it 💪

eshanrnh commented on 16/umbraco-cms/customizing/property-editors/property-value-converters.md:
```diff
@@ -1,64 +1,39 @@
 
-Example: Checking if the IPublishedPropertyType EditorAlias property is equal to the alias of the core content editor.\
+Example: Checking if the `IPublishedPropertyType.EditorAlias` property is equal to the alias of the core content editor.\
 This check is a string comparison but we recommend creating a constant for it to avoid spelling errors:
```
```suggestion
This check is a string comparison, but creating a constant is recommended to avoid spelling errors:
```

eshanrnh reviewed (approved):
Stepping in for Sofie while she's out sick. The PR looks great. 🙌 

I’ve added a few style guide suggestions and will take care of them. 

I’ve also updated the title in the `summary.md` file so it matches the article title.
